android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/FhirEngineExtension.kt

```diff
@@ -93,3 +96,29 @@ suspend fun FhirEngine.countUnSyncedResources() =
     .groupingBy { it.resourceType.spaceByUppercase() }
     .eachCount()
     .map { it.key to it.value }
+
+suspend fun <R : Resource> FhirEngine.batchedSearch(search: Search) =
+  if (search.count != null) {
+    this.search<R>(search)
+  } else {
+    val result = mutableListOf<SearchResult<R>>()
+    var offset = search.from ?: 0
+    val pageCount = 100
+    do {
+      search.from = offset
+      search.count = pageCount
+      val searchResults = this.search<R>(search)
+      result += searchResults
+      offset += searchResults.size
+    } while (searchResults.size == pageCount)
+
+    result
```

Comment:
Thanks for this @LZRS. 

I have a few questions and comments on this implementation. 

The decision to load this data in batches sounds reasonable, however, based on this implementation I believe the gain would be negligible or there would be no gain at all. Technically, the data will still be cached in memory inside the loop. This differs from the approach where we would load and process the data in batches. Data processing should happen after fetching each batch to improve the app responsiveness, otherwise, there are some drawbacks with batching including:

1. Each batch fetch requires a new database query, introducing overhead. The overhead of executing smaller queries may outweigh the benefit of lower memory.

2. Using `OFFSET` requires the database to skip over records 
before returning the batch. A larger offset would require scanning and discarding many rows thus if indexing is not done properly, this will impact fetching data in batches. 

@LZRS Would it be possible to measure the performance impact introduced by this implementation for comparison?

The paging3 library loads data in batches and allows for data processing with each fetch. 

cc @ndegwamartin @pld

Reply:
Yeah, generally, I think the performance gains would probably be negligible (in one of our cases, a difference of 20ms) especially when assuming only our query of interest would be running a single time. I think the significance of these changes would come about when we have multiple queries from different coroutines/threads.

Assuming for example we have a background long running operation accessing the db, and in the UI someone goes to a page that to render also requires a db fetch. Instead of waiting for the long running background operation to finish, the db fetch for the UI page can get queued in between the batches for the long running background. Looking to also add  integration tests on the same

The changes here would ensure that no single query(in batch) would take more than a couple of milliseconds, a long running query would still take long to be loaded (as it's split and cached in memory) but it would allow space for other queries to also get run.

I also agree that the best case would be data processing after fetching each batch although that would be quite tasking to implement, I feel.

In terms of MalawiCore, it improved the app's performance from [here](https://github.com/opensrp/fhircore/issues/3181#issuecomment-2134950554) to [here](https://github.com/opensrp/fhircore/issues/3181#issuecomment-2152451689).
We'll working on similar tests for FhirCore.

Use of paging3 to load would be the recommended approach but it would also need approval from the Google's sdk team

Note: 
> [Room](https://developer.android.com/reference/androidx/room/Transaction) will only perform at most one transaction at a time, additional transactions are queued and executed on a first come, first serve order.